src/Shouldly/Internals/EqualityComparerAdapter.cs

```diff
@@ -18,9 +18,9 @@ public EqualityComparerAdapter(IEqualityComparer<T> innerComparer)
             _innerComparer = innerComparer;
         }
 
-        public new bool Equals(object x, object y)
+        public new bool Equals(object? x, object? y)
         {
-            return _innerComparer.Equals((T) x, (T) y);
+            return _innerComparer.Equals((T)x!, (T)y!);
```

Comment:
The cast will always succeed when `T` is a nullable type at runtime (either a reference type or `Nullable<>`).